### PR TITLE
Improved the convergence control of the coupling iterations

### PR DESCRIPTION
**Pull request by wenqing**
as titled.

In addition, it includes several minor changes as
-  The commits in the closed PR of #1862 are merged to this PR
-  A new commit corrects the times of rejected steps and formats a warning message.

### CODE REVIEW

**Comment by wenqing** [comment]
The commits in the closed PR of #1862 are merged to this PR with two new TH tests:
27740470-6cacae44-5db2-11e7-9ced-139b745e2e15

**Comment by wenqing** [comment]
@waltherm 
The last commit corrects the times of rejected steps and formats a warning message. e.g.
```
warning: This step is rejected due to the relative change from the solution of the previous
         time step to the current solution exceeds the given tolerance of 0.01.
         This time step will be repeated with a new time step size of 2.46262e+08
         or the simulation will be halted.
warning: Time step 21 was rejected 2 times and it will be repeated with a reduced step size.
```

File: ProcessLib/UncoupledProcessesTimeLoop.cpp [hunk 1]
```diff
@@ -886,10 +886,17 @@ bool UncoupledProcessesTimeLoop::solveCoupledEquationSystemsByStaggeredScheme(
     bool coupling_iteration_converged = true;
     for (unsigned global_coupling_iteration = 0;
          global_coupling_iteration < _global_coupling_max_iterations;
-         global_coupling_iteration++)
+         global_coupling_iteration++, _global_coupling_conv_crit->reset())
     {
+        if (global_coupling_iteration == 0)
```
Comment (chleh) [inline ProcessLib/UncoupledProcessesTimeLoop.cpp]: Minor: I'd move that before the loop, maybe with condition
```
if (_global_coupling_max_iterations != 0)
```

Comment (wenqing) [inline ProcessLib/UncoupledProcessesTimeLoop.cpp]: Changed accordingly by amending the commit of `` [CPL] Improved the convergence ...``,  aada847.

File: ProcessLib/UncoupledProcessesTimeLoop.h [hunk 2]
```diff
@@ -67,7 +67,7 @@ class UncoupledProcessesTimeLoop
     std::vector<std::unique_ptr<SingleProcessData>> _per_process_data;
 
     bool _last_step_rejected = false;
-    bool _repeating_times_of_rejected_step = 0;
+    int _repeating_times_of_rejected_step = 0;
```
Comment (wenqing) [inline ProcessLib/UncoupledProcessesTimeLoop.h]: Changed to ``int``

File: ProcessLib/LiquidFlow/Tests.cmake [hunk 3]
```diff
@@ -111,6 +110,35 @@ AddTest(
...
 )
 
+AddTest(
+    NAME Adaptive_dt_ThermalConvectionFlow2D
+    PATH StaggeredCoupledProcesses/TH/ThermalConvection2D
+    EXECUTABLE ogs
+    EXECUTABLE_ARGS quad_5500x5500_adaptive_dt.prj
+    WRAPPER time
+    TESTER vtkdiff
+    REQUIREMENTS NOT OGS_USE_MPI
```
Comment (endJunction) [inline ProcessLib/LiquidFlow/Tests.cmake]: Did you run the tests using LIS? If not, please add `NOT OGS_USE_LIS`.

Comment (wenqing) [inline ProcessLib/LiquidFlow/Tests.cmake]: Added as
```
REQUIREMENTS NOT (OGS_USE_LIS OR OGS_USE_MPI)
```
by amending the commit of ``[dt] Added two more tests about about staggered scheme with adaptive ... ``.

**Comment by endJunction** [comment]
This is to be merged in few hours unless there is a review coming up....

**Pull request merged**